src/Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/TciOnDgGrid.cpp

```diff
@@ -0,0 +1,166 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/TciOnDgGrid.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/DgSubcell/PerssonTci.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/PalenzuelaEtAl.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace grmhd::ValenciaDivClean::subcell {
+template <typename RecoveryScheme>
+template <size_t ThermodynamicDim>
+bool TciOnDgGrid<RecoveryScheme>::apply(
+    const gsl::not_null<Variables<hydro::grmhd_tags<DataVector>>*> dg_prim_vars,
+    const Scalar<DataVector>& tilde_d, const Scalar<DataVector>& tilde_tau,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& tilde_s,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& tilde_b,
+    const Scalar<DataVector>& tilde_phi,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric,
+    const Scalar<DataVector>& sqrt_det_spatial_metric,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>& eos,
+    const Mesh<3>& dg_mesh, const TciOptions& tci_options,
+    const double persson_exponent) noexcept {
+  constexpr double persson_tci_epsilon = 1.0e-18;
+  const size_t number_of_points = dg_mesh.number_of_grid_points();
+  Variables<hydro::grmhd_tags<DataVector>> temp_prims(number_of_points);
+
+  // require: tilde_d/sqrt{gamma} >= 0.0 (or some positive user-specified value)
+  if (min(get(tilde_d) / get(sqrt_det_spatial_metric)) <
+      tci_options.minimum_rest_mass_density_times_lorentz_factor) {
+    return true;
+  }
+
+  // Check if we are in atmosphere (but not negative tildeD), and if so, then we
+  // continue using DG on this element.
+  if (max(get(tilde_d) / get(sqrt_det_spatial_metric) /
+          get(get<hydro::Tags::LorentzFactor<DataVector>>(*dg_prim_vars))) <
+          tci_options.atmosphere_density and
+      max(get(get<hydro::Tags::RestMassDensity<DataVector>>(*dg_prim_vars))) <
+          tci_options.atmosphere_density) {
+    // Even in atmosphere we need to recover the primitive magnetic field.
+    for (size_t i = 0; i < 3; ++i) {
+      get<hydro::Tags::MagneticField<DataVector, 3>>(*dg_prim_vars).get(i) =
+          tilde_b.get(i) / get(sqrt_det_spatial_metric);
+    }
+    get(get<hydro::Tags::DivergenceCleaningField<DataVector>>(*dg_prim_vars)) =
+        get(tilde_phi) / get(sqrt_det_spatial_metric);
+
+    return false;
```

Comment:
It may not be, we don't have enough data to know that. Until we do realistic simulations we won't know what the best TCI is. In general, as long as we are in atmosphere we are only evolving the magnetic field anyway (the velocity is evolved a bit with some changes that are in a different PR and will be added in another followup PR). The B field check in atmosphere can also incorrectly trigger atmosphere. It should really be handled by either an atmosphere fixer or a conservative variable fixer, but if we are in atmosphere there's little point in doing a recovery.